constraint/pkg/client/template_client.go

```diff
@@ -60,6 +63,44 @@ func (e *templateClient) ValidateConstraint(constraint *unstructured.Unstructure
 	return crds.ValidateCR(constraint, e.crd)
 }
 
+// ApplyDefaultParams will apply any default parameters defined in the CRD of the constraint's
+// corresponding template.
+// Assumes ValidateConstraint() called is called so the constraint is a valid CRD.
+func (e *templateClient) ApplyDefaultParams(constraint *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	cpy := constraint.DeepCopy()
+
+	params, found, err := unstructured.NestedFieldNoCopy(cpy.Object, "spec", "parameters")
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		params = make(map[string]interface{})
+	}
+
+	// sequentially get the structural schema for the spec field
```

Comment:
Any reason why we shouldn't apply all defaults specified by the CRD?

Reply:
( continuing our offline chat here)

> Any reason why we shouldn't apply all defaults specified by the CRD?

Not a particular reason but it seems that the other default we care about (legacy) works as expected so I was trying to be topical in only defaulting the OpenAPIV3 params.  

Without converting `e` or `e.crd` to a `apiextensions.JSONSchemaProps`, best we can do is :

```
	apiStructural, err := schema.NewStructural(e.crd.Spec.Validation.OpenAPIV3Schema)
	defaulting.Default(spec, apiStructural)

	if err := unstructured.SetNestedField(constraint.Object, spec, "spec"); err != nil {
		return nil, err
	}
```

Which, as discussed, doesn't work as one would think since it doesn't see a Default on the top level and bails early. 

---

Another take on this would be to look at how we default when we create the CRD:

https://github.com/open-policy-agent/frameworks/blob/3f237e2710faee505189afd605cfe75ed89dd446/constraint/pkg/client/crds/crds.go#L82-L85

I can look to explore that route but it may require some number of conversions IIUC, which adds more risk to this func. Also, I think it's worth having an "external" (as in external to the creation) func do this since `AddConstraint` can contain objects that are not constraints (which get weeded out on validation).

What do you think?

Reply:
`e.crd.Spec.Validation.OpenAPIV3Schema` describes the schema for the object root (not starting with spec).

To conform with the behavior of the K8s API server, let's apply the default to the root of the object. Doing this will make gator and `kubectl apply`-ing to a cluster behave identically (barring version skew WRT k8s defaulting code).

> Which, as discussed, doesn't work as one would think since it doesn't see a Default on the top level and bails early.

It works as K8s defaults work. Since we are piggybacking on that standard, following it is least surprising.

> Another take on this would be to look at how we default when we create the CRD

Not sure what you're getting at with this section. In any case, we shouldn't look to do too much in this PR... move slow, get feedback, iterate. There is more to lose by doing something that is broken than there is to gain by doing something quickly.

> Also, I think it's worth having an "external" (as in external to the creation) func

Given I didn't follow what you were saying, I may be missing something here, but we should not create an externally-accessible constraint defaulting function. I'm not sure what the use for this would be, and could be added at any time if it ever became necessary.

Reply:
> To conform with the behavior of the K8s API server, let's apply the default to the root of the object.

Okay I can buy that, in general, but, in this case, only `default`-ing to the root object will have the practical effect on not actually defaulting the parameters. So the UX is still, practically, broken. We can document the this papercut but I think it may be worth paying the price here to do the right thing.

>  Doing this will make gator and kubectl apply-ing to a cluster behave identically

This may be an aside but AFAICT, `kubectl apply` today won't default the parameters either (or maybe i'm holding it wrong).

Reply:
Per offline conversation, we're going to apply defaults at the root of the constraint object.